### PR TITLE
Fix spring resttemplate 3.0 support

### LINKED ISSUE #1519: NoClassDefFoundError: org/springframework/http/HttpRequest
This happens when using Spring RestTemplate 3.0.

In this version, the `org.springframework.http.HttpRequest` class is not available. It has been introduced in Spring 3.1.

We're depending on the `HttpRequest` class in `SpringRestRequestHeaderSette`r. The fix is quite simple as `HttpRequest`'s super class `HttpMessage` is available in 3.0 and declares the `getHeaders` method that we need.

### PR DESCRIPTION
**Pull request by SylvainJuge**
## What does this PR do?

Fixes #1519 - `ClassNotFoundError` thrown when using older spring-web versions and the spring-resttemplate plugin is active.

`ClientHttpRequest` only has a `#getURI` method after version 3.0.3, thus we can't support early `3.0.{0,1,2}` versions.
- [3.0.0](https://docs.spring.io/spring-framework/docs/3.0.0.RELEASE/javadoc-api/org/springframework/http/client/ClientHttpRequest.html)
- [3.0.3](https://docs.spring.io/spring-framework/docs/3.0.3.RELEASE/javadoc-api/org/springframework/http/client/ClientHttpRequest.html)

For the few unsupported versions, the span is not created and a debug message is logged.

This also adds integration tests for all the major/minor versions (but not the complete 140+ patch versions).
For those, a separate module was required in order to ensure that tested dependencies are in the isolated test classloader.

----

Update:
Support will be limited to versions `3.1.1+` as our instrumentation also relies on the [`getRawStatusCode`](https://docs.spring.io/spring-framework/docs/3.1.1.RELEASE/javadoc-api/org/springframework/http/client/ClientHttpResponse.html#getRawStatusCode()) that is not available before.

## Checklist

- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - ~~Added an API method or config option? Document in which version this will be introduced~~
  - [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: apm-agent-plugins/apm-spring-resttemplate/apm-spring-restemplate-plugin/src/main/java/co/elastic/apm/agent/resttemplate/SpringRestTemplateAdvice.java [hunk 1]
```diff
@@ -46,11 +47,22 @@
...
+        if (parent == null) {
             return null;
         }
-        final AbstractSpan<?> parent = TracerAwareInstrumentation.tracer.getActive();
-        Span span = HttpClientHelper.startHttpClientSpan(parent, Objects.toString(request.getMethod()), request.getURI(), request.getURI().getHost());
+        URI uri = null;
+        String host = "unknown";
+        try {
+            uri = request.getURI();
+        } catch (NoSuchMethodError ignored) {
```
Comment (felixbarny) [inline apm-agent-plugins/apm-spring-resttemplate/apm-spring-restemplate-plugin/src/main/java/co/elastic/apm/agent/resttemplate/SpringRestTemplateAdvice.java:59]: Better to make sure we don't instrument Spring RestTemplate < 3.0.3

File: docs/supported-technologies.asciidoc [hunk 2]
```diff
@@ -252,7 +252,7 @@ The spans are named after the schema `<method> <host>`, for example `GET elastic
 | 1.6.0
 
 |Spring RestTemplate
-|4+
+|3.0.6+
```
Comment (felixbarny) [inline docs/supported-technologies.asciidoc:255]:
```suggestion
|3.0.3+
```

File: apm-agent-plugins/apm-spring-resttemplate/apm-spring-restemplate-plugin/src/main/java/co/elastic/apm/agent/resttemplate/SpringRestTemplateAdvice.java (hunk 1 above)
Comment (SylvainJuge) [reply apm-agent-plugins/apm-spring-resttemplate/apm-spring-restemplate-plugin/src/main/java/co/elastic/apm/agent/resttemplate/SpringRestTemplateAdvice.java]: fixed in the last commit.

File: apm-agent-plugins/apm-spring-resttemplate/apm-spring-resttemplate-test/src/test/java/co.elastic.apm.agent.resttemplate/SpringRestTemplateVersionsIT.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class SpringRestTemplateVersionsIT {
+
+    @ParameterizedTest
+    @CsvSource({
+        "3.0.3.RELEASE", // lower versions are not supported
```
Comment (felixbarny) [inline apm-agent-plugins/apm-spring-resttemplate/apm-spring-resttemplate-test/src/test/java/co.elastic.apm.agent.resttemplate/SpringRestTemplateVersionsIT.java:39]: Did you manually test 3.0.0 to see that the agent no-ops and does not throw exceptions?

File: apm-agent-plugins/apm-spring-resttemplate/apm-spring-restemplate-plugin/pom.xml [hunk 4]
```diff
@@ -3,16 +3,16 @@
...
         <version>1.19.1-SNAPSHOT</version>
     </parent>
 
     <artifactId>apm-spring-resttemplate-plugin</artifactId>
     <name>${project.groupId}:${project.artifactId}</name>
 
     <properties>
-        <apm-agent-parent.base.dir>${project.basedir}/../..</apm-agent-parent.base.dir>
+        <apm-agent-parent.base.dir>${project.basedir}/../../..</apm-agent-parent.base.dir>
     </properties>
```
Comment (felixbarny) [inline apm-agent-plugins/apm-spring-resttemplate/apm-spring-restemplate-plugin/pom.xml:17]: How can we make sure we're always testing against the latest spring version as they get released? Maybe use a version range here or update the spring version via dependabot?

File: apm-agent-plugins/apm-spring-resttemplate/apm-spring-resttemplate-test/src/test/java/co.elastic.apm.agent.resttemplate/SpringRestTemplateVersionsIT.java (hunk 3 above)
Comment (SylvainJuge) [reply apm-agent-plugins/apm-spring-resttemplate/apm-spring-resttemplate-test/src/test/java/co.elastic.apm.agent.resttemplate/SpringRestTemplateVersionsIT.java]: This is now checked with integration test, there is no exception thrown nor visible when executing with unsupported versions.

File: apm-agent-plugins/apm-spring-resttemplate/apm-spring-restemplate-plugin/pom.xml (hunk 4 above)
Comment (SylvainJuge) [reply apm-agent-plugins/apm-spring-resttemplate/apm-spring-restemplate-plugin/pom.xml:17]: We can use [version ranges with Ivy](https://ant.apache.org/ivy/history/latest-milestone/ivyfile/dependency.html), I've changed the version 5.3.0 in the integration test to use that, and as a result it's currently testing with version 5.3.1.

**Review by felixbarny — APPROVED** [review]
👏

**Review by eyalkoren — CHANGES_REQUESTED** [review]
Awesome. Two required changes:
1. Fix module reference in pom (see inline suggestion)
2. `SpringRestTemplateVersionsIT.java` is located in a directory named `co.elastic.apm.agent.resttemplate` rather than a directory tree structure reflecting the package hierarchy.

**Review by eyalkoren — APPROVED** [review]

**Pull request merged**